### DISCUSSION TITLE
Visually Collapsing States

### DISCUSSION THREAD

**Discussion by adamfk** [StateSmith/StateSmith#202] [General]
Starting a discussion to explore the issue [iamjuing is having with visually collapsing states](https://github.com/StateSmith/StateSmith/issues/179#issuecomment-1552588494). Also how this compares to a new suggested feature of [$MACRO states](https://github.com/StateSmith/StateSmith/issues/179#issuecomment-1549629587).

> The problem with collapsing states though is that it's messing up the layout, while there still need to be enough space to be able to fully open them. Moreover, the composite states can grow over time. I find collapsing basically unusable here.

It is true that initially collapsing the state causes the shape size to change in draw.io, but that is easy enough to fix using the `Copy Size` and `Paste Size` draw.io feature. One day, the StateSmith plugin will do this by default.

![paste-size](https://github.com/StateSmith/StateSmith/assets/274012/973d42e3-15de-4963-88ba-043e6d88c8a7)

When my diagram grows large enough and I want to hide details, I collapse a state and leave it collapsed. I then rely on entry and exit points if I need something special for the visually collapsed state. 

> still need to be enough space to be able to fully open them.

@iamjuing, why do you need to fully open them once collapsed? You can enter a collapsed state by double clicking it or using the `Enter Group` button or shortcut. I've used this extensively for a number of large designs.

Thanks for helping me understand your use case better.

**Comment by iamjuing** [comment]
Please, keep in mind this has nothing to do with the $MACRO #201
It is just a (side)feature it provides ... allowing to extract complex parts of the diagram (and potentially re-use them if needed) and hide them behind the submachine state that has its own notation.

You made a good point though ... I don't really need to open them if they are collapsed. However, the thing is that I'm not really using the plugin as there is no way to have it permanently loaded in desktop draw.io. At least I am not aware of the way to do that. Also, the plugin is yet another dependency.

I think my vision at the moment is to have a complex diagram created from top-to-bottom where the sub-parts are spread over the draw.io page(s).

**Reply by adamfk** [reply]
Ah this makes sense. Drawio is a bit cumbersome for entering collapsed groups/states using the `Arrange` menu. 

The regular drawio keyboard shortcuts are helpful though:
* `CTRL+SHIFT+END` to enter a group
* `CTRL+SHIFT+HOME` to exit a group
* `SHIFT+HOME` to go home

A better solution is to "install" the StateSmith plugin into the draw.io offline application. It then only takes a few clicks to launch. I recently made a video showing how to do this: https://youtu.be/04EShKOB96c?t=241
